parlai/core/torch_agent.py

```diff
@@ -0,0 +1,258 @@
+
+        x_text = [ex['text_vec'] for ex in exs]
+        x_lens = [ex.shape[0] for ex in x_text]
+
+        if sort:
+            ind_sorted = sorted(range(len(x_lens)), key=lambda k: -x_lens[k])
+
+            exs = [exs[k] for k in ind_sorted]
+            valid_inds = [valid_inds[k] for k in ind_sorted]
+            x_text = [x_text[k] for k in ind_sorted]
+
+        padded_xs = torch.LongTensor(len(exs),
+                                     max(x_lens)).fill_(self.NULL_IDX)
+        if self.use_cuda:
+            padded_xs = padded_xs.cuda()
+
+        for i, ex in enumerate(x_text):
+            padded_xs[i, :ex.shape[0]] = ex
+
+        xs = padded_xs
+
+        eval_labels_avail = any(['eval_labels_vec' in ex for ex in exs])
+        labels_avail = any(['labels_vec' in ex for ex in exs])
+        some_labels_avail = eval_labels_avail or labels_avail
+
+        # set up the target tensors
+        ys = None
+        labels = None
+        if some_labels_avail:
+            # randomly select one of the labels to update on (if multiple)
+            if labels_avail:
+                field = 'labels'
+            else:
+                field = 'eval_labels'
+
+            num_choices = [len(ex.get(field + "_vec", [])) for ex in exs]
+            choices = [random.choice(range(num)) if num != 0 else -1
+                       for num in num_choices]
+            label_vecs = [ex[field + "_vec"][choices[i]]
+                          if choices[i] != -1 else torch.LongTensor([])
+                          for i, ex in enumerate(exs)]
+            labels = [ex[field][choices[i]]
+                      if choices[i] != -1 else ''
+                      for i, ex in enumerate(exs)]
+            y_lens = [y.shape[0] for y in label_vecs]
+            padded_ys = torch.LongTensor(len(exs),
+                                         max(y_lens)).fill_(self.NULL_IDX)
+            if self.use_cuda:
+                padded_ys = padded_ys.cuda()
+            for i, y in enumerate(label_vecs):
+                if y.shape[0] != 0:
+                    padded_ys[i, :y.shape[0]] = y
+            ys = padded_ys
+
+        return Batch(xs, ys, labels, valid_inds)
+
+    def unmap_valid(self, predictions, valid_inds, batch_size):
+        """Re-order permuted predictions to the initial ordering, includes the
+        empty observations.
+
+        :param predictions: output of module's predict function
+        :param valid_inds: original indices of the predictions
+        :param batch_size: overall original size of batch
+        """
+        unpermuted = [None]*batch_size
+        for pred, idx in zip(predictions, valid_inds):
+            unpermuted[idx] = pred
+        return unpermuted
+
+    def maintain_dialog_history(self, observation, reply='',
+                                useStartEndIndices=False,
+                                splitSentences=False):
+        """Keeps track of dialog history, up to a truncation length.
```

Comment:
might want to also list the :params here as well (just to keep in line with the previous functions in this class)